Tests/iaas/flavor-naming/flavor-add-extra-specs.py

```diff
@@ -0,0 +1,345 @@
+        else:
+            flavor.delete_extra_specs_property(compute, prop)
+            if prop in flavor.extra_specs:
+                del flavor.extra_specs[prop]
+        return 0
+    except openstack.exceptions.SDKException as exc:
+        print(f"ERROR: Could not set {prop} for {flavor.name}: {exc!r}",
+              file=sys.stderr)
+        return 1
+
+
+def check_extra_type(flavor, prop, val, dct):
+    """Check extra_specs['scs:prop'] for flavor
+    It should be set and consistent with val, translated with dct.
+    Returns 1 is the extra_spec needs to change"""
+    spec = f"scs:{prop}"
+    if val:
+        expected = dct[val]
+    else:
+        expected = None
+    if spec in flavor.extra_specs:
+        setting = flavor.extra_specs[spec]
+        if setting != expected:
+            print(f"ERROR: flavor {flavor.name} has {spec} set to {setting}, expect {expected}",
+                  file=sys.stderr)
+        else:
+            return 0
+    elif not expected:
+        return 0
+    flavor.extra_specs[spec] = expected
+    if not QUIET:
+        print(f"INFO  {flavor.name}: Update extra_spec {spec} to {flavor.extra_specs[spec]}")
+    return 1
+
+
+def main(argv):
+    "Entry point"
+    global DEBUG, QUIET, NOCHANGE
+    errors = 0
+    cloud = None
+    disk0_type = None
+    cpu_type = None
+
+    if "OS_CLOUD" in os.environ:
+        cloud = os.environ["OS_CLOUD"]
+    try:
+        opts, flvs = getopt.gnu_getopt(argv, "hdqt:p:c:n",
+                                       ("help", "debug", "quiet", "disk0-type=",
+                                        "cpu-type=", "os-cloud=", "no-change"))
+    except getopt.GetoptError as exc:
+        print(f"CRITICAL: {exc!r}", file=sys.stderr)
+        usage(1)
+    for opt in opts:
+        if opt[0] == "-h" or opt[0] == "--help":
+            usage(0)
+        if opt[0] == "-d" or opt[0] == "--debug":
+            DEBUG = True
+        if opt[0] == "-q" or opt[0] == "--quiet":
+            QUIET = True
+        if opt[0] == "-n" or opt[0] == "--no-change":
+            NOCHANGE = True
+        if opt[0] == "-c" or opt[0] == "--os-cloud":
+            cloud = opt[1]
+        if opt[0] == "-t" or opt[0] == "--disk0-type":
+            disk0_type = opt[1]
+            if disk0_type not in DISKTYPE_KEY:
+                disk0_type = revert_dict(disk0_type, DISKTYPE_KEY)
+                if not disk0_type:
+                    return 2
+        if opt[0] == "-p" or opt[0] == "--cpu-type":
+            cpu_type = opt[1]
+            if cpu_type not in CPUTYPE_KEY:
+                cpu_type = revert_dict(cpu_type, CPUTYPE_KEY)
+                if not cpu_type:
+                    return 2
+
+    if not cloud:
+        print("ERROR: Need to pass -c|--os-cloud|OS_CLOUD env", file=sys.stderr)
+        usage(3)
+
+    conn = openstack.connect(cloud)
+    conn.authorize()
+    compute = conn.compute
+
+    flavors = compute.flavors()
+    for flavor in flavors:
```

Comment:
insert a separate step before this loop that filters `flavors` depending on whether all SCS flavors should be processed or a fixed list of flavors supplied via name by the user

Reply:
We need flavor instances, not just a list of name.
So something like this would please?
```python
def filter_flavors(conn, specified):
  flavors = []
  for flv in conn.compute.flavors():
    if specified:
      if flv.name in specified:
        flavors.append(flv)
    else:
      if flv.name[0:4] == "SCS-":
        flavors.append(flv)
  return flavors
```

Reply:
I thought more along the lines of:

```python
    if flvs:
        flavors = [flv for flv in flavors if flv.name in flvs]
    else:
        flavors = [flv for flv in flavors if flv.name.startswith('SCS-')
```

or maybe

```python
    predicate = (lambda fn: fn in flvs) if flvs else (lambda fn: fn.startswith('SCS-')
    flavors = [flv for flv in flavors if predicate(flv.name)]
```

Reply:
Just FYI, I think the second variant could also be written

```python
    predicate = flvs.__contains__ if flvs else (lambda fn: fn.startswith('SCS-'))
    flavors = [flv for flv in flavors if predicate(flv.name)]
```

Reply:
If you are into functional programming, you could do something like (untested!):
```python
class matcher:
  def __init__(self, lst):
    self.lst = lst
  def flv_match(self, flavor):
    return (self.lst and flavor.name in self.lst) or (not self.lst and flavor.name[0:4] == "SCS-")

def filter_flavors(conn, specified):
  mymatch = matcher(specified)
  return filter(mymatch.flv_match, conn.compute.flavors())
```
More elegant, less readable ...

Reply:
Surprisingly, this works. Wrap a `list()` around it for testing ...

Reply:
Well, I had to objectives:

- pull the if out of the loop
- get by with 2 to 4 lines, because more is not warranted for this simple task

So far, each of your variants has the if inside the loop, and it has 9 to 10 lines ;)